From 87a6cbbe668edffb8f1ddd612e33fbd84c4a0ef2 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@freescale.com>
Date: Tue, 15 Sep 2015 16:47:50 +0800
Subject: [PATCH 0742/1691] MLK-11514 [MX6SX] No GUI show and the display is
 blank after bootup to x rootfs with LVDS.100%.

commit 87a6cbbe668edffb8f1ddd612e33fbd84c4a0ef2 from
git://git.freescale.com/imx/linux-2.6-imx.git imx_4.1.15_1.0.0_ga

The reason for this issue is that, on imx6sx platform, there is a display mix
to control the display modules(including lcdif) power on/off. When the fb
driver calls the 'pm_runtime_put_sync_suspend()', the display mix will be
powered down. And at this moment, all the register settings for lcdif always
have no effects. So the pm_runtime calling points in mxsfb driver should be
optimized to avoid this kind of cases.

Signed-off-by: Fancy Fang <chen.fang@freescale.com>
(cherry picked from commit 3b85eb989c89fb1c6d08215e435d72ab175e4e1e)

Conflicts:
	drivers/video/mxsfb.c
---
 drivers/video/fbdev/mxsfb.c | 19 +++++++++++--------
 1 file changed, 11 insertions(+), 8 deletions(-)

diff --git a/drivers/video/fbdev/mxsfb.c b/drivers/video/fbdev/mxsfb.c
index 7f40946..25ef8ed 100644
--- a/drivers/video/fbdev/mxsfb.c
+++ b/drivers/video/fbdev/mxsfb.c
@@ -529,8 +529,6 @@ static void mxsfb_enable_controller(struct fb_info *fb_info)
 		}
 	}
 
-	pm_runtime_get_sync(&host->pdev->dev);
-
 	/* the pixel clock should be disabled before
 	 * trying to set its clock rate successfully.
 	 */
@@ -610,8 +608,6 @@ static void mxsfb_disable_controller(struct fb_info *fb_info)
 	reg = readl(host->base + LCDC_VDCTRL4);
 	writel(reg & ~VDCTRL4_SYNC_SIGNALS_ON, host->base + LCDC_VDCTRL4);
 
-	pm_runtime_put_sync_suspend(&host->pdev->dev);
-
 	host->enabled = 0;
 
 	if (host->reg_lcd) {
@@ -898,8 +894,12 @@ static int mxsfb_blank(int blank, struct fb_info *fb_info)
 	case FB_BLANK_VSYNC_SUSPEND:
 	case FB_BLANK_HSYNC_SUSPEND:
 	case FB_BLANK_NORMAL:
-		if (host->enabled)
+		if (host->enabled) {
 			mxsfb_disable_controller(fb_info);
+			pm_runtime_put_noidle(&host->pdev->dev);
+			pm_runtime_put_sync_suspend(&host->pdev->dev);
+			pm_runtime_get_noresume(&host->pdev->dev);
+		}
 
 		clk_disable_disp_axi(host);
 		clk_disable_axi(host);
@@ -909,10 +909,13 @@ static int mxsfb_blank(int blank, struct fb_info *fb_info)
 	case FB_BLANK_UNBLANK:
 		fb_info->var.activate = (fb_info->var.activate & ~FB_ACTIVATE_MASK) |
 				FB_ACTIVATE_NOW | FB_ACTIVATE_FORCE;
+
+		clk_enable_pix(host);
+		clk_enable_axi(host);
+		clk_enable_disp_axi(host);
+
 		if (!host->enabled) {
-			clk_enable_pix(host);
-			clk_enable_axi(host);
-			clk_enable_disp_axi(host);
+			pm_runtime_get_sync(&host->pdev->dev);
 
 			writel(0, host->base + LCDC_CTRL);
 			mxsfb_set_par(&host->fb_info);
-- 
1.9.1

